source/extensions/common/tap/tap.h

```diff
@@ -0,0 +1,124 @@
+ */
+class ExtensionConfig {
+public:
+  virtual ~ExtensionConfig() = default;
+
+  /**
+   * @return the ID to use for admin extension configuration tracking (if applicable).
+   */
+  virtual const std::string& adminId() PURE;
+
+  /**
+   * Clear any active tap configuration.
+   */
+  virtual void clearTapConfig() PURE;
+
+  /**
+   * Install a new tap configuration.
+   * @param proto_config supplies the generic tap config to install. Not all configuration fields
+   *        may be applicable to an extension (e.g. HTTP fields). The extension is free to fail
+   *        the configuration load via exception if it wishes.
+   * @param admin_streamer supplies the singleton admin sink to use for output if the configuration
+   *        specifies that output type. May not be used if the configuration does not specify
+   *        admin output. May be nullptr if admin is not used to supply the config.
+   */
+  virtual void newTapConfig(envoy::service::tap::v2alpha::TapConfig&& proto_config,
+                            Sink* admin_streamer) PURE;
+};
+
+class Matcher;
+using MatcherPtr = std::unique_ptr<Matcher>;
+
+/**
+ * Base class for all tap matchers.
+ *
+ * A high level note on the design of tap matching which is different from other matching in Envoy
+ * due to a requirement to support streaming matching (match as new data arrives versus
+ * calculating the match given all available data at once).
+ * - The matching system is composed of a constant matching configuration. This is essentially
+ *   a tree of matchers given logical AND, OR, NOT, etc.
+ * - A per-stream/request matching status must be kept in order to compute interim match status.
+ * - In order to make this computationally efficient, the matching tree is kept in a vector, with
+ *   all references to other matchers implemented using an index into the vector.
+ * - The previous point allows the creation of a per-stream/request vector of booleans of the same
+ *   size as the matcher vector. Then, when match status is updated given new information, the
+ *   vector of booleans can be easily updated using the same indexes as in the constant match
+ *   configuration.
+ * - Finally, a matches() function can be trivially implemented by looking in the status vector at
+ *   the index position that the current matcher is located in.
+ */
+class Matcher {
+public:
+  virtual ~Matcher() = default;
+
+  /**
+   * @return the matcher's index in the match tree vector (see above).
+   */
+  size_t index() { return my_index_; }
+
+  /**
+   * Update match status given new information.
+   * @param request_headers supplies the request headers, if available.
+   * @param response_headers supplies the response headers, if available.
+   * @param statuses supplies the per-stream-request match status vector which must be the same
+   *                 size as the match tree vector (see above).
+   */
+  virtual bool updateMatchStatus(const Http::HeaderMap* request_headers,
```

Comment:
There's something a bit weird here to the casual reader about the mix of pure interface and concrete implementation. We have very low level details, e.g. tree representation, in this interface class while still leaving some methods PURE. How many plausible implementations are there of this interface given this?

Reply:
There are quite a few implementations, but I will just move to tap_matcher.h with the concrete implementations.